### PR TITLE
fixed issue #11051

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Fix Github issue #11051 

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [ ] Bug-Fix for a *released version* (did you remember to port this to all relevant release branches?)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

#### Related Information

- [x] There is a GitHub Issue reported by a Community User: #11051 

### Testing & Verification

This PR adds tests that were used to verify all changes:

- [x] Added new **integration tests** (i.e. in shell_server / shell_server_aql)

http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/8454/

### CODE REVIEW

File: arangod/Aql/IndexNodeOptimizerRules.cpp [hunk 1]
```diff
@@ -119,6 +120,14 @@ void arangodb::aql::lateDocumentMaterializationRule(Optimizer* opt,
       if (!indexNode->canApplyLateDocumentMaterializationRule() || indexNode->isLateMaterialized()) {
         continue; // loop is already optimized
       }
+      auto producingNode = dynamic_cast<DocumentProducingNode const*>(loop);
```
Comment (iurii-i-popov) [inline arangod/Aql/IndexNodeOptimizerRules.cpp]: Why do you use a cast here? You can check indexNode. It is better to use ExecutionNode::castTo<>().

Comment (jsteemann) [reply arangod/Aql/IndexNodeOptimizerRules.cpp]: True. Thanks!

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/8462/

**Comment by jsteemann** [comment]
tests blue

**Pull request merged**